sklearn/neighbors/_dist_metrics.pyx

```diff
@@ -986,25 +998,21 @@ cdef class HaversineDistance(DistanceMetric):
        D(x, y) = 2\\arcsin[\\sqrt{\\sin^2((x1 - y1) / 2)
                                 + \\cos(x1)\\cos(y1)\\sin^2((x2 - y2) / 2)}]
     """
+
+    def _validate_data(self, X):
+        if X.shape[1] != 2:
+            raise ValueError("Haversine distance only valid "
+                             "in 2 dimensions")
+
     cdef inline DTYPE_t rdist(self, DTYPE_t* x1, DTYPE_t* x2,
                               ITYPE_t size) nogil except -1:
-        if size != 2:
-            with gil:
-                raise ValueError("Haversine distance only valid "
-                                 "in 2 dimensions")
         cdef DTYPE_t sin_0 = sin(0.5 * (x1[0] - x2[0]))
         cdef DTYPE_t sin_1 = sin(0.5 * (x1[1] - x2[1]))
         return (sin_0 * sin_0 + cos(x1[0]) * cos(x2[0]) * sin_1 * sin_1)
 
     cdef inline DTYPE_t dist(self, DTYPE_t* x1, DTYPE_t* x2,
-                              ITYPE_t size) nogil except -1:
-        if size != 2:
-            with gil:
-                raise ValueError("Haversine distance only valid in 2 dimensions")
-        cdef DTYPE_t sin_0 = sin(0.5 * (x1[0] - x2[0]))
-        cdef DTYPE_t sin_1 = sin(0.5 * (x1[1] - x2[1]))
-        return 2 * asin(sqrt(sin_0 * sin_0
-                             + cos(x1[0]) * cos(x2[0]) * sin_1 * sin_1))
+                             ITYPE_t size) nogil except -1:
+        return 2 * asin(sqrt(self.rdist(x1, x2, size)))
```

Comment:
Not a blocker:

I have always wondered if `inline` works for methods such as `rdist`. My guess is that the `inline` is being ignored. Is this true @scoder ?

Reply:
Look at the generated C code to find out. :)

Reply:
In this case, the function is not `CYTHON_INLINE`. But in `PyFuncDistance._dist` is `CYTHON_INLINE`.

I have seen in the [cython docs](https://cython.readthedocs.io/en/latest/src/userguide/pyrex_differences.html?highlight=inline#cdef-inline) that this would be ignored in class-level cdefs.

Reply:
The question is not so much if the function/method is declared as `CYTHON_INLINE` in the code, but how it is called. If it's called via its internal C function name, then it's (practically) inlined, because the C compiler can see which function is being called and optimise the call by tying it into the calling code. Subclass overrides are then also not taken into account (as intended). For this, Cython has to know at compile time which exact method is being called, i.e. of which exact class.

If, in the C code, the method is called by hopping through `vtab` pointers instead, then the C compiler can not see the function target and will make an indirect call to whatever method implementation is found in the cdef method table at runtime.

Reply:
That makes sense. Thank you for your insight!